mdbook-tera-backend/src/tera_renderer/custom_component.rs

```diff
@@ -0,0 +1,133 @@
+use serde::Deserialize;
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::path::{Path, PathBuf};
+
+use tera::Tera;
+
+use crate::{RendererError, Result};
+
+use super::RenderingContext;
+
+fn make_strip_prefix_function() -> impl tera::Function {
+    move |args: &HashMap<String, serde_json::value::Value>| -> tera::Result<tera::Value> {
+        let string = args
+            .get("s")
+            .ok_or_else(|| tera::Error::from(format!("No s argument provided")))?
+            .as_str()
+            .ok_or_else(|| tera::Error::from(format!("S has invalid type, expected string")))?;
+        let prefix = args
+            .get("prefix")
+            .ok_or_else(|| tera::Error::from(format!("No prefix argument provided")))?
+            .as_str()
+            .ok_or_else(|| {
+                tera::Error::from(format!("Prefix has invalid type, expected string"))
+            })?;
+        string
+            .strip_prefix(prefix)
+            .map(|s| tera::Value::String(s.to_owned()))
+            .ok_or_else(|| tera::Error::from(format!("Could not strip prefix")))
+    }
+}
+
+pub struct CustomComponent {
+    template: Tera,
+    name: String,
+    /// Used to generate unique ids for each component to prevent collisions in javascript with query selectors.
+    counter: RefCell<u32>,
+}
+
+impl CustomComponent {
+    pub fn new(name: &str, template_str: &str, dependencies: &[&Self]) -> Result<CustomComponent> {
+        let mut template = Tera::default();
+        for dep in dependencies {
+            template.extend(&dep.template)?;
+        }
+        template.add_raw_template(name, template_str)?;
+        template.register_function("strip_prefix", make_strip_prefix_function());
```

Comment:
This is the thing which others cannot do when they use our tool.

Reply:
This one is not needed because there is already a builtin for this. The other method is a nice helper for our specific use case. In the end, I don't think it is unreasonable that people will want us to add new methods. But tera already has a lot of builtins (even `now` and `date` for instance)

Reply:
> In the end, I don't think it is unreasonable that people will want us to add new methods. But tera already has a lot of builtins (even `now` and `date` for instance)

Yes and no... the problem for me is ownership and maintenance. If someone needs a `fetch_stock_quote` helper function, should we add it? We'll have to maintain it going forward and I don't really want to do that.

Delegating everything to the template system makes the decision easy: we surface relevant data and then people can manipulate it as much as they want using functions/macros in the template system.

Reply:
Yes, I agree and I can't think of a use case yet for a domain specific method